### PR TITLE
Initial apparmor profile for qemu

### LINKED ISSUE #7181: Apparmor profile for qemu
It would be nice to have `dnsmasq` confined by an Apparmor profile.

Here is the profile I've been using since using lxd from the snap:

```
# Author: Simon Deziel <simon@sdeziel.info>
#include <tunables/global>

profile snap-lxd-dnsmasq /snap/lxd/[0-9]*/bin/dnsmasq flags=(attach_disconnected) {
  #include <abstractions/base>
  #include <abstractions/dbus>
  #include <abstractions/nameservice>

  capability chown,
  capability net_bind_service,
  capability setgid,
  capability setuid,
  capability dac_override,
  capability net_admin,         # for DHCP server
  capability net_raw,           # for DHCP server ping checks
  network inet raw,
  network inet6 raw,

  /var/snap/lxd/common/lxd/networks/*/dnsmasq.{leases,pid,raw} rw,
  /var/snap/lxd/common/lxd/networks/*/dnsmasq.hosts/{,*} rw,

  @{PROC}/sys/net/ipv6/conf/*/mtu r,
  /var/lib/snapd/hostfs/etc/group              r,
  /var/lib/snapd/hostfs/etc/host.conf          r,
  /var/lib/snapd/hostfs/etc/hosts              r,
  /var/lib/snapd/hostfs/etc/nsswitch.conf      r,
  /var/lib/snapd/hostfs/etc/gai.conf           r,
  /var/lib/snapd/hostfs/etc/passwd             r,
  /var/lib/snapd/hostfs/etc/protocols          r,
  /var/lib/snapd/hostfs/etc/resolv.conf        r,

  /var/lib/snapd/hostfs/run/{resolvconf,NetworkManager,systemd/resolve,connman}/resolv.conf r,
  /var/lib/snapd/hostfs/etc/resolvconf/run/resolv.conf r,
  /var/lib/snapd/hostfs/run/systemd/resolve/stub-resolv.conf r,

  /snap/lxd/[0-9]*/lib/**.so*                  mr,
  /run/dnsmasq.pid                             w,

  # Site-specific additions and overrides. See local/README for details.
  #include <local/snap.lxd.dnsmasq>
}
```

Since I started using it, it only required 1 modification recently (allow `mr` on the .so provided by the snap) so it should be fairly low maintenance for the lxd team.

### PR DESCRIPTION
**Pull request by stgraber**
Closes #7181

### CODE REVIEW

File: lxd/instance/drivers/driver_qemu.go [hunk 1]
```diff
@@ -504,6 +506,12 @@ func (vm *qemu) Freeze() error {
 
 // onStop is run when the instance stops.
 func (vm *qemu) onStop(target string) error {
+	ctxMap := log.Ctx{
```
Comment (tomponline) [inline lxd/instance/drivers/driver_qemu.go:509]: @stgraber not a blocker now but would be good to add a common logger property like we do for networks, devices and storage in the future.

File: lxd/instance/drivers/driver_qemu.go [hunk 2]
```diff
@@ -4389,6 +4430,11 @@ func (vm *qemu) LogPath() string {
 	return shared.LogPath(name)
 }
 
+// EarlyLogFilePath returns the instance's early log path.
+func (vm *qemu) EarlyLogFilePath() string {
```
Comment (tomponline) [inline lxd/instance/drivers/driver_qemu.go:4434]: @stgraber does this need to be exported, I cant see it being used outside of the package?

**Comment by tomponline** [comment]
@stgraber looks like they maybe an issue with this, I just tried it locally and I can't VMs now (I'm not running from within the snap). 

```
lxc launch images:ubuntu/focal v1 --vm -s btrfs1
Creating v1
Starting v1                                 
Error: Failed to run: forklimits limit=memlock:unlimited:unlimited -- /usr/bin/qemu-system-x86_64 -S -name v1 -uuid 1c018703-255b-45db-b27b-6d025e336a08 -daemonize -cpu host -nographic -serial chardev:console -nodefaults -no-reboot -no-user-config -sandbox on,obsolete=deny,elevateprivileges=allow,spawn=deny,resourcecontrol=deny -readconfig /var/log/lxd/v1/qemu.conf -pidfile /var/log/lxd/v1/qemu.pid -D /var/log/lxd/v1/qemu.log -chroot /var/lib/lxd/virtual-machines/v1 -smbios type=2,manufacturer=Canonical Ltd.,product=LXD -runas nobody: : Process exited with a non-zero value
Try `lxc info --show-log local:v1` for more info
```

```
lxc info --show-log local:v1
Name: v1
Location: none
Remote: unix://
Architecture: x86_64
Created: 2020/09/16 07:58 UTC
Status: Stopped
Type: virtual-machine
Profiles: default

Log:

qemu-system-x86_64:/var/log/lxd/v1/qemu.conf:23: Failed to create PTY: No such file or directory
```

```
dmesg | grep apparmor
[ 1763.172913] audit: type=1400 audit(1600243153.278:7994): apparmor="STATUS" operation="profile_load" profile="unconfined" name="lxd-v1_</var/lib/lxd>" pid=9500 comm="apparmor_parser"
[ 1763.285220] audit: type=1400 audit(1600243153.390:7995): apparmor="DENIED" operation="open" profile="lxd-v1_</var/lib/lxd>" name="/home/user/go/src/github.com/lxc/lxd/" pid=9515 comm="qemu-system-x86" requested_mask="r" denied_mask="r" fsuid=0 ouid=1000
[ 1763.285224] audit: type=1400 audit(1600243153.390:7996): apparmor="DENIED" operation="open" profile="lxd-v1_</var/lib/lxd>" name="/sys/bus/nd/devices/" pid=9515 comm="qemu-system-x86" requested_mask="r" denied_mask="r" fsuid=0 ouid=0
[ 1763.287475] audit: type=1400 audit(1600243153.390:7997): apparmor="DENIED" operation="open" profile="lxd-v1_</var/lib/lxd>" name="/sys/devices/system/node/" pid=9515 comm="qemu-system-x86" requested_mask="r" denied_mask="r" fsuid=0 ouid=0
[ 1763.330674] audit: type=1400 audit(1600243153.434:7998): apparmor="DENIED" operation="open" profile="lxd-v1_</var/lib/lxd>" name="/dev/ptmx" pid=9525 comm="qemu-system-x86" requested_mask="r" denied_mask="r" fsuid=0 ouid=0
```

**Comment by tomponline** [comment]
I've fixed some of the issues with this:

```
diff --git a/lxd/apparmor/instance_qemu.go b/lxd/apparmor/instance_qemu.go
index 7ceb02dc5..96df4ae56 100644
--- a/lxd/apparmor/instance_qemu.go
+++ b/lxd/apparmor/instance_qemu.go
@@ -20,14 +20,15 @@ profile "{{ .name }}" flags=(attach_disconnected,mediate_deleted) {
   # Needed by qemu
   /{,usr/}bin/qemu*                         mrix,
   /dev/hugepages/**                         w,
-  /dev/kvm                                  w,
-  /dev/net/tun                              w,
-  /dev/ptmx                                 w,
+  /dev/kvm                                  rw,
+  /dev/net/tun                              rw,
+  /dev/ptmx                                 rw,
   /dev/vfio/**                              w,
-  /dev/vhost-net                            w,
-  /dev/vhost-vsock                          w,
+  /dev/vhost-net                            rw,
+  /dev/vhost-vsock                          rw,
   /etc/ceph/**                              r,
   /usr/share/OVMF/OVMF_CODE.fd              kr,
+  /usr/share/qemu/kvmvapic.bin              r,
   owner @{PROC}/@{pid}/task/@{tid}/comm     rw,
```

However its now complaining about the instance specific path `/var/lib/lxd/storage-pools/btrfs1/virtual-machines/v1/config/`

```
[ 2349.352675] audit: type=1400 audit(1600243739.461:8033): apparmor="DENIED" operation="open" profile="lxd-v1_</var/lib/lxd>" name="/var/lib/lxd/storage-pools/btrfs1/virtual-machines/v1/config/" pid=12990 comm="qemu-system-x86" requested_mask="r" denied_mask="r" fsuid=0 ouid=65534
```

Before I go much further I'll wait to discuss with you as to best approach as not familiar with apparmor config.

**Comment by stgraber** [comment]
Interesting. I only tested with the snap, so looks like the distro qemu is behaving differently.

**Comment by tomponline** [comment]
@stgraber more failures after latest patch:

```
qemu-system-x86_64:/var/log/lxd/v1/qemu.conf:224: failed to find romfile "vgabios-virtio.bin"
```

```
[25941.001453] audit: type=1400 audit(1600267331.284:8572): apparmor="DENIED" operation="open" profile="lxd-v1_</var/lib/lxd>" name="/sys/bus/nd/devices/" pid=204198 comm="qemu-system-x86" requested_mask="r" denied_mask="r" fsuid=0 ouid=0
[25941.001595] audit: type=1400 audit(1600267331.284:8573): apparmor="DENIED" operation="open" profile="lxd-v1_</var/lib/lxd>" name="/sys/devices/system/node/" pid=204198 comm="qemu-system-x86" requested_mask="r" denied_mask="r" fsuid=0 ouid=0
[25941.013499] audit: type=1400 audit(1600267331.296:8574): apparmor="DENIED" operation="open" profile="lxd-v1_</var/lib/lxd>" name="/sys/module/vhost/parameters/max_mem_regions" pid=204212 comm="qemu-system-x86" requested_mask="r" denied_mask="r" fsuid=0 ouid=0
[25941.024314] audit: type=1400 audit(1600267331.304:8575): apparmor="DENIED" operation="open" profile="lxd-v1_</var/lib/lxd>" name="/sys/module/vhost/parameters/max_mem_regions" pid=204212 comm="qemu-system-x86" requested_mask="r" denied_mask="r" fsuid=0 ouid=0
[25941.029179] audit: type=1400 audit(1600267331.312:8576): apparmor="DENIED" operation="open" profile="lxd-v1_</var/lib/lxd>" name="/usr/share/seabios/vgabios-virtio.bin" pid=204212 comm="qemu-system-x86" requested_mask="r" denied_mask="r" fsuid=0 ouid=0
```

**Comment by tomponline** [comment]
https://github.com/lxc/lxd/pull/7878 fixes these issues.

**Pull request merged**